rust/src/storage/s3/dynamodb_lock.rs

```diff
@@ -439,8 +450,17 @@ struct AcquireLockState<'a> {
 }
 
 impl<'a> AcquireLockState<'a> {
+    /// If lock is expirable (lease_duration is set) then this function returns `true`
+    /// if the elapsed time sine `started` is reached `timeout_in`.
     fn has_timed_out(&self) -> bool {
-        self.started.elapsed() > self.timeout_in
+        self.started.elapsed() > self.timeout_in && {
+            let non_expirable = if let Some(ref cached_lock) = self.cached_lock {
+                cached_lock.lease_duration.is_none()
+            } else {
+                false
+            };
+            !non_expirable
```

Comment:
I would avoid the double negative here to make the code more readable:

```suggestion
            let expirable = if let Some(ref cached_lock) = self.cached_lock {
                cached_lock.lease_duration.is_some()
            } else {
                true
            };
            expirable
```